### PR TITLE
Upgradable L2 contract

### LINKED ISSUE #159: Add L2 upgradability

### PR DESCRIPTION
**Pull request by DefiCake**
Adds support for upgradability to the bridge contract

Closes #159

### CODE REVIEW

File: packages/fungible-token/bridge-fungible-token/implementation/src/main.sw [hunk 1]
```diff
@@ -169,8 +170,9 @@ impl Bridge for Contract {
         });
     }
 
+    #[storage(read)]
     fn bridged_token_gateway() -> b256 {
-        BRIDGED_TOKEN_GATEWAY
+        storage.l1_gateway.try_read().unwrap_or(BRIDGED_TOKEN_GATEWAY)
```
Comment (luizstacio) [inline packages/fungible-token/bridge-fungible-token/implementation/src/main.sw:175]: Why this was moved to be a storage?

Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/implementation/src/main.sw:175]: I cannot recall if this was Swayam or me, but this is not set in stone yet.

Depending on how you initialize the contract's parameters, you might want to have these in configurables (constants) or in storage. You basically do not want your contract to be callable during the initialisation stage.

In that sense I think we are just following EVM mechanics here. We will have to review the whole design once it's finished. Especially because the final proxy to implement was the simple one, not diamonds - which is how I began this PR.

Fwiw I think we are going to move this one back to configurables. Again, nothing is final yet.

File: packages/base-asset/src/main.sw [hunk 2]
```diff
@@ -66,7 +66,7 @@ impl SRC20 for Contract {
     /// }
     /// ```
     #[storage(read)]
-    fn total_supply(asset: AssetId) -> Option<u64> {
+    fn total_supply(_asset: AssetId) -> Option<u64> {
```
Comment (bitzoic) [inline packages/base-asset/src/main.sw:69]: This change should be removed. Any changes to the base asset contract need to be done in a separate PR

File: packages/fungible-token/bridge-fungible-token/implementation/src/main.sw [hunk 3]
```diff
@@ -62,7 +61,7 @@ use utils::{
     encode_data,
     encode_register_calldata,
 };
-use src_20::SRC20;
+use standards::src20::SRC20;
```
Comment (bitzoic) [inline packages/fungible-token/bridge-fungible-token/implementation/src/main.sw:64]: Was this change not done in https://github.com/FuelLabs/fuel-bridge/pull/201?

File: packages/fungible-token/bridge-fungible-token/implementation/src/main.sw [hunk 4]
```diff
@@ -182,6 +184,10 @@ impl Bridge for Contract {
     fn asset_to_l1_address(asset_id: AssetId) -> b256 {
         _asset_to_l1_address(asset_id)
     }
+
+    fn double_value(foo: u64) -> u64 {
```
Comment (bitzoic) [inline packages/fungible-token/bridge-fungible-token/implementation/src/main.sw:188]: What is this for?

File: packages/fungible-token/bridge-fungible-token/interface/src/bridge.sw [hunk 5]
```diff
@@ -35,4 +36,6 @@ abi Bridge {
 
     #[storage(read)]
     fn asset_to_l1_address(asset_id: AssetId) -> b256;
+
+    fn double_value(foo: u64) -> u64;
```
Comment (bitzoic) [inline packages/fungible-token/bridge-fungible-token/interface/src/bridge.sw:40]: Why is this needed?

File: packages/fungible-token/bridge-fungible-token/proxy/Forc.toml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+[project]
+authors = ["Fuel Labs <contact@fuel.sh>"]
+entry = "proxy.sw"
+license = "Apache-2.0"
+name = "proxy"
+
+[dependencies]
+contract_message_receiver = { path = "../../../message-predicates/contract-message-receiver" }
+interface = { path = "../interface" }
+standards = { git = "https://github.com/FuelLabs/sway-standards", branch = "master" }
```
Comment (bitzoic) [inline packages/fungible-token/bridge-fungible-token/proxy/Forc.toml:10]:
```suggestion
standards = { git = "https://github.com/FuelLabs/sway-standards", tag = "v0.4.4" }
```

File: packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw [hunk 7]
```diff
@@ -0,0 +1,153 @@
+contract;
+
+use std::{
+    call_frames::first_param,
+    constants::ZERO_B256,
```
Comment (bitzoic) [inline packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw:5]:
```suggestion
```

File: packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    execution::run_external,
+    hash::Hash,
+    hash::sha256,
```
Comment (bitzoic) [inline packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw:8]:
```suggestion
    hash::{sha256, Hash},
```

File: packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+use interface::{bridge::Bridge, src7::{Metadata, SRC7}};
+
+#[namespace(SRC14)]
+storage {
+    // target is at sha256("storage_SRC14_0")
+    target: ContractId = ContractId::zero(),
+    owner: State = State::Initialized(Identity::Address(Address::zero())),
+}
+
+impl SRC5 for Contract {
```
Comment (bitzoic) [inline packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw:22]: This will need to be changed see: https://github.com/FuelLabs/sway-standards/pull/107

File: packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+    call_frames::first_param,
+    constants::ZERO_B256,
+    execution::run_external,
+    hash::Hash,
+    hash::sha256,
+    string::String,
+};
+use standards::{src14::SRC14, src20::SRC20, src5::{AccessError, SRC5, State}};
+use contract_message_receiver::MessageReceiver;
+use interface::{bridge::Bridge, src7::{Metadata, SRC7}};
```
Comment (bitzoic) [inline packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw:13]: Why is src7 imported from interface and not standards?

File: packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+
+impl ChangeOwner for Contract {
+    #[storage(read, write)]
+    fn change_owner(new_owner: Identity) {
+        only_owner();
+        storage.owner.write(State::Initialized(new_owner));
+    }
+}
+
+// abi Diamonds {
```
Comment (bitzoic) [inline packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw:67]: Why is this commented out code here?

File: packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+
+// impl MessageReceiver for Contract {
+//     #[payable]
+//     #[storage(read, write)]
+//     fn process_message(msg_idx: u64) {
+//         run_external(TARGET)
+//     }
+// }
+
+// // Uncomment this and compiler fails
```
Comment (bitzoic) [inline packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw:126]: Is this still true?

File: packages/fungible-token/bridge-fungible-token/tests/functions/bridge/mod.rs [hunk 13]
```diff
@@ -401,22 +409,14 @@ mod success {
...
-    async fn check_bridged_token_gateway() {
-        let contract = create_token().await;
-
-        let response = bridged_token_gateway(&contract).await;
-
-        assert_eq!(
-            response,
-            Bits256(*Address::from_str(BRIDGED_TOKEN_GATEWAY).unwrap())
+        let supply = total_supply(
+            &implementation_contractid,
```
Comment (bitzoic) [inline packages/fungible-token/bridge-fungible-token/tests/functions/bridge/mod.rs:413]: Shouldn't this be going through the proxy contract?

**Review by bitzoic — COMMENTED** [review]
Looks like the proxy contract isn't tested and all tests go through the implementation contract.

File: packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw (hunk 12 above)
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw]: Removed

File: packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw (hunk 11 above)
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw]: Removed

File: packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw (hunk 10 above)
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw:13]: Cannot remember the exact reason except that it had to do with the pending PR https://github.com/FuelLabs/fuel-bridge/pull/103, something about having compilation or SDK problems due to the encoding. It has been a long standing task to implement SRC-8 since the new encoding is available now

File: packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw (hunk 9 above)
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw]: Changed

File: packages/fungible-token/bridge-fungible-token/interface/src/bridge.sw (hunk 5 above)
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/interface/src/bridge.sw:40]: A debugging function back from when LDC was still semi functional. Removed

File: packages/fungible-token/bridge-fungible-token/implementation/src/main.sw (hunk 3 above)
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/implementation/src/main.sw:64]: Yeah I am not sure what is exactly flagging this as a change. Probably because the file has moved and because of the multiple rebases / merges that have happened since this PR started, git is not able to exactly track what happened here.

File: packages/fungible-token/bridge-fungible-token/implementation/src/main.sw (hunk 1 above)
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/implementation/src/main.sw:175]: Removed

File: packages/base-asset/src/main.sw (hunk 2 above)
Comment (DefiCake) [reply packages/base-asset/src/main.sw]: Solved

File: packages/fungible-token/bridge-fungible-token/tests/functions/bridge/mod.rs (hunk 13 above)
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/tests/functions/bridge/mod.rs:418]: It is going through the proxy, if you look one line below, but the SDK needs the implementation contract to add it to the inputs of the call.

File: packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    #[storage(read)]
+    fn _proxy_target() -> ContractId {
+        storage.target.read()
+    }
+
+    #[storage(read, write)]
+    fn _proxy_change_owner(new_owner: Identity) {
+        only_owner();
```
Comment (K1-R1) [inline packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw:71]: As the `storage.owner` starts as `State::Uninitialized`; a new owner can never be set as this `only_owner()` will always revert.
Instead either:
- `storage.owner` can start as `State::Initialized` as in: https://github.com/FuelLabs/sway-standards/blob/master/examples/src5-ownership/initialized_example/src/initialized_example.sw
- an `initialize_ownership` method can be included as in: https://github.com/FuelLabs/sway-libs/blob/0f47d33d6e5da25f782fc117d4be15b7b12d291b/libs/src/ownership.sw#L127

**Review by K1-R1 — CHANGES_REQUESTED** [review]
Requesting changes as it seems the proxy's owner can never be set

File: packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw (hunk 14 above)
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw:71]: @SwayStar123 wrote the storage initialization and I packed it up here:

https://github.com/FuelLabs/fuel-bridge/blob/ae2ec304463e2fd5f903440d8fc562346148d551/packages/fungible-token/bridge-fungible-token/tests/utils/setup.rs#L590

I am not sure if there is a best practice defined. I would have gone the storage.unwrap_or(configurable) path if I had implemented it because I didn't know you can set storage slots on deployment.

Comment (SwayStar123) [reply packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw:71]: I believe i set it so the owner is set at deployment (with the storage slots changed via the sdk). If this is not best practice feel free to change it.
It is not possible to use a configurable in storage hence why i did this

**Comment by DefiCake** [comment]
As agreed in internal talks:
- Use of configurables instead of custom storage slots for initialization of ownership
- Input sanitization to avoid Address::default() as parameter for the change owner function

File: packages/fungible-token/bridge-fungible-token/tests/functions/message_receiver/mod.rs [hunk 15]
```diff
@@ -48,20 +50,23 @@ mod success {
...
             BRIDGED_TOKEN_DECIMALS,
-            Some(configurables.clone()),
+            proxy_id,
             false,
             None,
         )
         .await;
 
-        let (bridge, utxo_inputs) = setup_environment(
+        let (implementation_contractid, bridge, utxo_inputs) = setup_environment(
```
Comment (K1-R1) [inline packages/fungible-token/bridge-fungible-token/tests/functions/message_receiver/mod.rs:69]:
```suggestion
        let (implementation_contract_id, bridge, utxo_inputs) = setup_environment(
```
only a nit, feel free to ignore if it would delay

File: packages/fungible-token/bridge-fungible-token/tests/functions/proxy/mod.rs [hunk 16]
```diff
@@ -0,0 +1,374 @@
...
+                if address == wallet.address().clone().into()
+            ),
+            "Ownership was not initialized or owner is not the expected address"
+        );
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn proxy_set_target() -> anyhow::Result<()> {
```
Comment (K1-R1) [inline packages/fungible-token/bridge-fungible-token/tests/functions/proxy/mod.rs:283]: Can you add an only_owner version of this test/ check in this test that a non-owner calling the method errors

File: packages/fungible-token/bridge-fungible-token/tests/utils/builder.rs [hunk 17]
```diff
@@ -41,13 +41,13 @@ pub async fn build_contract_message_tx(
...
     }
-
     // Start building tx list of outputs
     tx_outputs.push(Output::contract(1u16, Bytes32::zeroed(), Bytes32::zeroed()));
+    tx_outputs.push(Output::contract(2u16, Bytes32::zeroed(), Bytes32::zeroed()));
 
-    // If there is more than 1 contract input, it means this is a deposit to contract.
-    if number_of_contracts > 1usize {
-        tx_outputs.push(Output::contract(2u16, Bytes32::zeroed(), Bytes32::zeroed()));
+    // If there is more than 2 contract input, it means this is a deposit to contract.
```
Comment (K1-R1) [inline packages/fungible-token/bridge-fungible-token/tests/utils/builder.rs:48]:
```suggestion
    // If there is more than 2 contract inputs, it means this is a deposit to a contract.
```

**Review by K1-R1 — COMMENTED** [review]
Overall looks good to me; I prefer the use of configureables. Happy to approve once the test for `only_owner` is added to the test as per the comment below

File: packages/fungible-token/bridge-fungible-token/tests/functions/proxy/mod.rs (hunk 16 above)
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/tests/functions/proxy/mod.rs:338]: Done

File: packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+    #[storage(read, write)]
+    fn _proxy_change_owner(new_owner: Identity);
+
+    #[storage(read, write)]
+    fn _proxy_revoke_ownership();
+}
+
+configurable {
+    INITIAL_OWNER: State = State::Uninitialized,
+    INITIAL_TARGET: ContractId = ContractId::from(ZERO_B256)
```
Comment (bitzoic) [inline packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw:34]:
```suggestion
    INITIAL_TARGET: ContractId = ContractId::zero()
```

File: packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+    target: Option<ContractId> = None,
+    // owner is at sha256("storage_SRC14_1")
+    owner: State = State::Uninitialized,
+}
+
+impl SRC14 for Contract {
+    #[storage(write)]
+    fn set_proxy_target(new_target: ContractId) {
+        only_owner();
+        require(new_target.bits() != ZERO_B256, ProxyErrors::IdentityZero);
```
Comment (bitzoic) [inline packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw:49]:
```suggestion
        require(new_target.bits() != b256::zero(), ProxyErrors::IdentityZero);
```

File: packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw [hunk 20]
```diff
@@ -0,0 +1,104 @@
+contract;
+
+use std::{
+    call_frames::first_param,
+    execution::run_external,
+    hash::{sha256, Hash},
+    string::String,
+    constants::ZERO_B256,
```
Comment (bitzoic) [inline packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw:8]:
```suggestion
```

File: packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+
+    #[storage(read)]
+    fn _proxy_target() -> ContractId {
+        storage.target.read().unwrap_or(INITIAL_TARGET)
+    }
+
+    #[storage(read, write)]
+    fn _proxy_change_owner(new_owner: Identity) {
+        only_owner();
+        require(new_owner.bits() != ZERO_B256, ProxyErrors::IdentityZero);
```
Comment (bitzoic) [inline packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw:95]:
```suggestion
        require(new_owner.bits() != b256::zero(), ProxyErrors::IdentityZero);
```

**Review by bitzoic — APPROVED** [review]
Just some minor nits

File: packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw [hunk 22]
```diff
@@ -0,0 +1,104 @@
+contract;
+
+use std::{
+    call_frames::first_param,
+    execution::run_external,
+    hash::{sha256, Hash},
```
Comment (bitzoic) [inline packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw:6]: I don't believe this import is used

Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw]: Done

**Review by K1-R1 — APPROVED** [review]

File: packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+    fn _proxy_owner() -> State;
+
+    #[storage(read)]
+    fn _proxy_target() -> ContractId;
+
+    #[storage(read, write)]
+    fn _proxy_change_owner(new_owner: Identity);
+
+    #[storage(read, write)]
+    fn _proxy_revoke_ownership();
```
Comment (bitzoic) [inline packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw:21]: Have you considered a single function for ownership changes rather than two separate functions? `State::Revoked` would be passed as the function argument.
```sway
    #[storage(read, write)]
    fn _proxy_change_owner(new_owner: State);
```

**Review by bitzoic — APPROVED** [review]

File: packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw (hunk 23 above)
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/proxy/src/proxy.sw:21]: I have, but I thought it would be more idiomatic this way and somewhat harder to mess it up, idk
